Mathlib/GroupTheory/BHNeumannLemma.lean

```diff
@@ -0,0 +1,188 @@
+      exact Set.subset_univ _
+    -- Thus `y • C ⊆ ⋃ k ∈ {k : H k ≠ C}, (y * x⁻¹ * g k) • H k` for all `y : G`, that is,
+    -- every left coset of `C` is contained in a finite union of left cosets of the
+    -- subgroups `H k ≠ C`.
+    replace this : ∀ z ∈ x • (C : Set G), ∃ k, ¬H k = C ∧ z ∈ g k • (H k : Set G) := by
+      simpa [Set.subset_def] using this
+    replace this (y : G) :
+        y • (C : Set G) ⊆
+          ⋃ i ∈ Finset.univ.filter (¬H · = C), (y * x⁻¹ * g i) • (H i : Set G) := by
+      suffices ∀ z ∈ y • (C : Set G), ∃ i, ¬H i = C ∧ z ∈ (y * x⁻¹ * g i) • (H i : Set G) by
+        simpa [Set.subset_def] using this
+      intro z hz
+      rw [mem_leftCoset_iff] at hz
+      obtain ⟨k, hk, h⟩ := this (x * y⁻¹ * z) <| by
+        rwa [mem_leftCoset_iff, ← mul_assoc, ← mul_assoc, mul_left_inv, one_mul]
+      refine ⟨k, hk, ?_⟩
+      rwa [mem_leftCoset_iff, mul_inv_rev, mul_inv_rev, inv_inv, mul_assoc, ← mem_leftCoset_iff]
+    -- Then `G` can also be covered by a finite union of left cosets of the subgroups `H k ≠ C`.
+    let κ₁ := { i // i ∈ Finset.univ.filter (H · ≠ C) }
+    let κ₂ := { i // i ∈ Finset.univ.filter (H · = C) }
+    let κ := Sum (κ₂ × κ₁) κ₁
+    have hκ₁ (i : κ₁) : (H i : Subgroup G) ∈ s.erase C := by
+      rw [Finset.mem_erase, ← Subtype.ext_iff.ne]
+      exact ⟨(Finset.mem_filter.mp i.property).right, Finset.coe_mem _⟩
+    let f : κ → G | Sum.inl ⟨k, i⟩ => g k * x⁻¹ * g i | Sum.inr i => g i
+    let K : κ → { x // x ∈ s.erase ↑C } | Sum.inl ⟨_, i⟩ => ⟨H i, hκ₁ i⟩ | Sum.inr i => ⟨H i, hκ₁ i⟩
+    -- By the induction hypothesis, one of the subgroups `H k ≠ C` has finite index.
+    specialize ih κ ((Set.range K).toFinset.map ⟨_, Subtype.val_injective⟩) <| by
+      rw [Finset.card_map]
+      refine le_trans ?_ ((Finset.card_erase_of_mem C.property).le.trans (by omega))
+      rw [← Finset.card_attach (s := s.erase C), Finset.attach_eq_univ]
+      exact Finset.card_le_card (Finset.subset_univ _)
+    specialize ih (fun k => ⟨Set.rangeFactorization K k,
+      Finset.mem_map.mpr ⟨K k, Set.mem_toFinset.mpr ⟨k, rfl⟩, rfl⟩⟩)
+    specialize ih <| by
+      intro ⟨x, hx⟩
+      have ⟨y, hy, hyx⟩ := Finset.mem_map.mp hx
+      have ⟨k, hk⟩ := Set.mem_toFinset.mp hy
+      rw [Function.Embedding.coeFn_mk] at hyx
+      exact ⟨k, by simp only [hk, hyx, Set.rangeFactorization_coe]⟩
+    specialize ih f <| by
+      rw [Set.iUnion_eq_univ_iff]
+      intro y
+      cases (Set.mem_union _ _ _).mp (hG.symm.subset (Set.mem_univ y)) with
+      | inr hy =>
+        rw [Set.mem_iUnion₂] at hy
+        obtain ⟨i, hi, hy⟩ := hy
+        exact ⟨Sum.inr ⟨i, hi⟩, hy⟩
+      | inl hy =>
+        rw [Set.mem_iUnion₂] at hy
+        obtain ⟨k, hk, hy⟩ := hy
+        have hk' : H k = C := by simpa using hk
+        specialize this (g k) (hk' ▸ hy)
+        rw [Set.mem_iUnion₂] at this
+        obtain ⟨i, hi, hy⟩ := this
+        have : κ := Sum.inl ⟨⟨k, hk⟩, ⟨i, hi⟩⟩
```

Comment:
```suggestion
```
I haven't looked at the rest of the code, but hopefully this will fix the CI failure.

Reply:
Thanks!